### PR TITLE
Change napi_callback to return napi_value instead of void

### PR DESCRIPTION
**Pull request by boingoing**
Change ```napi_callback``` to return ```napi_value``` instead of requiring the callback to manually call ```napi_set_return_value```.

When we invoke the callback, we will check the return value and call ```SetReturnValue``` ourselves. If the callback returns ```NULL```, we don't set the return value in v8 which would have the same effect as previously if the callback didn't call ```napi_set_return_value```. Seems to be a more natural way to handle return values from callbacks. As a consequence, remove ```napi_set_return_value```.

Add a ```napi_value``` to ```napi_property_descriptor``` to support string values which couldn't be passed in the ```utf8name``` parameter or symbols as property names. Class names, however, cannot be symbols so this ```napi_value``` must be a string type in that case.

Remove all of the ```napi_callback_info``` helpers except for ```napi_get_cb_info``` and make all the parameters to ```napi_get_cb_info``` optional except for argc.

Update all the test collateral according to these changes.

### CODE REVIEW

**Comment by boingoing** [comment]
@digitalinfinity @jasongin 

Please take a look, I'll push a further commit to fix lint bugs found via CI.

**Comment by boingoing** [comment]
Passing CI:
https://ci.nodejs.org/job/node-test-commit-linux-abi/93/

File: src/node_api.cc
Comment (jasongin) [inline src/node_api.cc]: If the property name is not a string, I think this should just skip setting the name of the function, not return an error.

Comment (jasongin) [reply src/node_api.cc]: Actually I don't think `SetClassName()` should be called at all here. The V8 doc for that API says:

> This is used for printing objects created with the function created from the FunctionTemplate as its constructor. 

What this was intended to do was set the function (class method) name. It's odd that `v8::FunctionTemplate` doesn't have a `SetName()` method, while `v8::Function` does.

Comment (jasongin) [reply src/node_api.cc]: Probably the function will get its name from the property name automatically, so there is no need to set the name separately.

**Review by jasongin — CHANGES_REQUESTED** [review]
Looks good aside from the one issue about class method names. I do think these are definitely improvements to the API surface. And eliminating `napi_set_return_value` will improve performance a little.

But we should consider whether to hold this until after the node PR lands. Does anyone else have thoughts on that?

**Comment by boingoing** [comment]
Sounds right. I was also confused about why we were calling SetClassName here. Seems that is for constructors, not methods.

File: src/node_api.cc
Comment (boingoing) [reply src/node_api.cc]: Sounds right. I was also confused about why we were calling ```SetClassName()``` here. Seems that is for constructors, not methods.

**Comment by jasongin** [comment]
> all the parameters to napi_get_cb_info optional except for argc

Why isn't `argc` optional also? Maybe methods which don't expect (or ignore) any parameters shouldn't need to supply that.

**Comment by boingoing** [comment]
This is a little disruptive, yes, waiting for the PR to land is probably fine. I don't want it to get stale, though, since there are so many invasive changes to the test code. Most of the change is in the test code, which is probably not getting as much attention over in the PR.

**Comment by boingoing** [comment]
> Why isn't argc optional also? Maybe methods which don't expect (or ignore) any parameters shouldn't need to supply that.

I thought about it but figured it didn't make sense because argc is both an in and an out parameter so it's required if argv is provided. That might be fine, though, to say if argv is provided that argc becomes non-optional.

File: src/node_api.cc
Comment (jasongin) [inline src/node_api.cc]: Anna just commented on the main PR that node style prefers `if (value != nullptr)`. Also below.

**Comment by jasongin** [comment]
> argv is provided that argc becomes non-optional

Yes. While updating leveldown code I encountered several places where it would be more convenient to pass nullptr for both argc and argv.

File: src/node_api.cc
Comment (boingoing) [reply src/node_api.cc]: Saw that and updated uses in this PR.

**Comment by boingoing** [comment]
Another CI job after rebasing
https://ci.nodejs.org/job/node-test-commit-linux-abi/96/

**Comment by mhdawson** [comment]
landed in main repo closing.

**Pull request closed without merging**
